File: ios/Spectra/PBKDF2Module.swift
/*
 * Copyright (c) 2026 MOZAGA FOUNDATION.
 * SPDX-License-Identifier: AGPL-3.0-only OR LicenseRef-Spectra-Commercial
 * See LICENSE.md, LICENSE-AGPL-3.0.txt, and LICENSE-COMMERCIAL.md for details.
 */

import Foundation
import CommonCrypto

@objc(PBKDF2Module)
class PBKDF2Module: NSObject {
  @objc
  func deriveKey(
    _ pin: String,
    salt: String,
    iterations: NSNumber,
    resolver resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    DispatchQueue.global(qos: .userInitiated).async {
      guard let saltData = Data(base64Encoded: salt) else {
        reject("ERR_SALT", "Invalid PBKDF2 salt", nil)
        return
      }

      let iterationCount = iterations.intValue
      guard iterationCount > 0 else {
        reject("ERR_ITERATIONS", "PBKDF2 iterations must be greater than zero", nil)
        return
      }

      guard let pinData = pin.data(using: .utf8), !pinData.isEmpty else {
        reject("ERR_PIN", "PIN must not be empty", nil)
        return
      }

      let derivedKeyLength = 32
      var derivedKey = Data(count: derivedKeyLength)
      let status = derivedKey.withUnsafeMutableBytes { derivedBytes in
        saltData.withUnsafeBytes { saltBytes in
          pinData.withUnsafeBytes { pinBytes in
            CCKeyDerivationPBKDF(
              CCPBKDFAlgorithm(kCCPBKDF2),
              pinBytes.bindMemory(to: Int8.self).baseAddress,
              pinData.count,
              saltBytes.bindMemory(to: UInt8.self).baseAddress,
              saltData.count,
              CCPseudoRandomAlgorithm(kCCPRFHmacAlgSHA256),
              UInt32(iterationCount),
              derivedBytes.bindMemory(to: UInt8.self).baseAddress,
              derivedKeyLength
            )
          }
        }
      }

      guard status == kCCSuccess else {
        reject("ERR_DERIVE", "PBKDF2 failed with status \(status)", nil)
        return
      }

      resolve(derivedKey.base64EncodedString())
    }
  }

  @objc
  static func requiresMainQueueSetup() -> Bool { false }
}

File: android/app/src/main/java/com/mozaga/exo/PBKDF2Module.kt
/*
 * Copyright (c) 2026 MOZAGA FOUNDATION.
 * SPDX-License-Identifier: AGPL-3.0-only OR LicenseRef-Spectra-Commercial
 * See LICENSE.md, LICENSE-AGPL-3.0.txt, and LICENSE-COMMERCIAL.md for details.
 */

package com.mozaga.exo

import android.util.Base64
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import java.util.Arrays
import java.util.concurrent.Executors
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

class PBKDF2Module(reactContext: ReactApplicationContext) :
    ReactContextBaseJavaModule(reactContext) {

    companion object {
        private const val KEY_LENGTH_BITS = 256
        private const val MAX_ITERATIONS = 2_000_000
    }

    private val executor = Executors.newSingleThreadExecutor()

    override fun getName(): String = "PBKDF2Module"

    @ReactMethod
    fun deriveKey(pin: String, salt: String, iterations: Int, promise: Promise) {
        executor.execute {
            try {
                if (pin.isEmpty()) {
                    promise.reject("ERR_PIN", "PIN must not be empty")
                    return@execute
                }
                if (iterations <= 0 || iterations > MAX_ITERATIONS) {
                    promise.reject("ERR_ITERATIONS", "Invalid PBKDF2 iteration count")
                    return@execute
                }

                val saltBytes = Base64.decode(salt, Base64.DEFAULT)
                val passwordBytes = pin.toByteArray(Charsets.UTF_8)
                val key = pbkdf2HmacSha256(passwordBytes, saltBytes, iterations, KEY_LENGTH_BITS / 8)
                Arrays.fill(passwordBytes, 0)
                promise.resolve(Base64.encodeToString(key, Base64.NO_WRAP))
            } catch (error: Throwable) {
                promise.reject("ERR_DERIVE", "PBKDF2 failed", error)
            }
        }
    }

    private fun pbkdf2HmacSha256(
        password: ByteArray,
        salt: ByteArray,
        iterations: Int,
        dkLen: Int
    ): ByteArray {
        val mac = Mac.getInstance("HmacSHA256")
        mac.init(SecretKeySpec(password, "HmacSHA256"))
        val hLen = mac.macLength
        val blockCount = (dkLen + hLen - 1) / hLen
        val derived = ByteArray(dkLen)
        var offset = 0

        for (blockIndex in 1..blockCount) {
            val block = salt + byteArrayOf(
                (blockIndex ushr 24).toByte(),
                (blockIndex ushr 16).toByte(),
                (blockIndex ushr 8).toByte(),
                blockIndex.toByte()
            )
            var u = mac.doFinal(block)
            val t = u.copyOf()

            for (round in 2..iterations) {
                u = mac.doFinal(u)
                for (index in t.indices) {
                    t[index] = (t[index].toInt() xor u[index].toInt()).toByte()
                }
            }

            val bytesToCopy = minOf(hLen, dkLen - offset)
            System.arraycopy(t, 0, derived, offset, bytesToCopy)
            offset += bytesToCopy
        }

        return derived
    }
}
